lib/Sema/TypeCheckAttr.cpp

```diff
@@ -2374,6 +2288,11 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
                   diag::differentiable_attr_specified_not_function,
                   primalSpecifier.Name, /*isPrimal*/ true);
     };
+    std::function<void()> primalInvalidTypeContextDiagnostic = [&]() {
+      TC.diagnose(primalNameLoc,
+                  diag::differentiable_attr_function_not_same_type_context,
+                  primalSpecifier.Name);
```

Comment:
Do you need a `setInvalid` here?

Reply:
I think `setInvalid` is handled by the block below:
```c++
primal = TC.lookupFuncDecl(...)
if (!primal) {
  attr->setInvalid();
  return;
}
```

If any diagnostic is emitted, `primal` will be `nullptr` and `attr->setInvalid()` will execute.